test/jasmine/tests/calcdata_test.js

```diff
@@ -28,4 +29,215 @@ describe('calculated data and points', function() {
+                    categorymode: 'category ascending',
+                    categorylist: ['b','a','d','e','c'] // These must be ignored. Alternative: error?
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(11);
+                expect(gd.calcdata[0][1].y).toEqual(13);
+                expect(gd.calcdata[0][2].y).toEqual(15);
+                expect(gd.calcdata[0][3].y).toEqual(14);
+                expect(gd.calcdata[0][4].y).toEqual(12);
+            });
+
+            it('should output categories in ascending domain alphanumerical order, excluding undefined', function() {
+
+                Plotly.plot(gd, [{x: ['c',undefined,'e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'category ascending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(11);
+                expect(gd.calcdata[0][1].y).toEqual(15);
+                expect(gd.calcdata[0][2].y).toEqual(14);
+                expect(gd.calcdata[0][3].y).toEqual(12);
+            });
+        });
+
+        describe('codomain numerical category ordering', function() {
+
+            it('should output categories in ascending codomain numerical order', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'value ascending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(11);
+                expect(gd.calcdata[0][1].y).toEqual(12);
+                expect(gd.calcdata[0][2].y).toEqual(13);
+                expect(gd.calcdata[0][3].y).toEqual(14);
+                expect(gd.calcdata[0][4].y).toEqual(15);
+            });
+
+            it('should output categories in descending codomain numerical order', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'value descending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(15);
+                expect(gd.calcdata[0][1].y).toEqual(14);
+                expect(gd.calcdata[0][2].y).toEqual(13);
+                expect(gd.calcdata[0][3].y).toEqual(12);
+                expect(gd.calcdata[0][4].y).toEqual(11);
+            });
+
+            it('should output categories in descending codomain numerical order, excluding nulls', function() {
+
+                Plotly.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,null,13,14]}], { xaxis: {
+                    type: 'category',
+                    categorymode: 'value descending'
+                }});
+
+                expect(gd.calcdata[0][0].y).toEqual(15);
+                expect(gd.calcdata[0][1].y).toEqual(14);
+                expect(gd.calcdata[0][2].y).toEqual(12);
+                expect(gd.calcdata[0][3].y).toEqual(11);
```

Comment:
this is correct, the `null` items should disappear from the calcdata data. Similar to the current [behavior](http://codepen.io/etpinard/pen/NxZLvZ).